File: tools/cmux-home/go/internal/home/state_test.go
package home

import "testing"

func TestParseStateAcceptsSharedLikeShape(t *testing.T) {
	state, err := ParseState([]byte(`{
		"generatedAt": "2026-05-11T20:00:00Z",
		"sessions": [
			{
				"id": "row-1",
				"session_id": "claude-native-1",
				"agent": {"kind": "Claude Code"},
				"state": "in_progress",
				"prompt": "Fix the reload link",
				"workingDirectory": "/tmp/cmux",
				"git": {"branch": "feat-home"},
				"messages": [{"role": "user", "content": "ignored because prompt wins"}]
			},
			{
				"sessionId": "codex-native-1",
				"kind": "codex",
				"status": "completed",
				"messages": [{"role": "user", "content": "Review adapter metadata"}]
			}
		],
		"tasks": [
			{"id": "task-1", "title": "Start home", "status": "pending", "adapter": "Open Code"}
		]
	}`))
	if err != nil {
		t.Fatalf("ParseState returned error: %v", err)
	}

	if got := len(state.Sessions); got != 2 {
		t.Fatalf("sessions = %d, want 2", got)
	}
	first := state.Sessions[0]
	if first.Adapter != "claude" {
		t.Fatalf("first adapter = %q, want claude", first.Adapter)
	}
	if first.Status != "working" {
		t.Fatalf("first status = %q, want working", first.Status)
	}
	if first.SessionID != "claude-native-1" {
		t.Fatalf("first session id = %q", first.SessionID)
	}
	if first.Branch != "feat-home" {
		t.Fatalf("first branch = %q", first.Branch)
	}
	if got := state.Sessions[1].Title; got != "Review adapter metadata" {
		t.Fatalf("second title = %q", got)
	}
	if got := state.Tasks[0].Adapter; got != "opencode" {
		t.Fatalf("task adapter = %q, want opencode", got)
	}
}

func TestFallbackStateHasAllPrototypeAdapters(t *testing.T) {
	state := FallbackState()
	counts := AdapterCounts(state.Sessions)
	got := map[string]int{}
	for _, count := range counts {
		got[count.Adapter] = count.Count
	}
	for _, adapter := range []string{"claude", "codex", "opencode", "pi"} {
		if got[adapter] != 1 {
			t.Fatalf("fallback count for %s = %d, want 1", adapter, got[adapter])
		}
	}
}

func TestParseStateRejectsUnsupportedSchemaVersion(t *testing.T) {
	_, err := ParseState([]byte(`{"schemaVersion":2,"sessions":[]}`))
	if err == nil {
		t.Fatal("ParseState returned nil error for unsupported schemaVersion")
	}
}

func TestParseStateRejectsSchemaStatusOutsideContract(t *testing.T) {
	_, err := ParseState([]byte(`{
		"schemaVersion": 1,
		"sessions": [
			{"id": "bad", "agent": "codex", "agentSessionId": "bad", "status": "failed"}
		]
	}`))
	if err == nil {
		t.Fatal("ParseState returned nil error for invalid schema status")
	}
}

func TestParseStateReadsCanonicalSchemaFields(t *testing.T) {
	state, err := ParseState([]byte(`{
		"schemaVersion": 1,
		"sessions": [
			{
				"id": "codex:canonical",
				"agent": "codex",
				"agentSessionId": "canonical",
				"title": "Canonical",
				"status": "awaiting",
				"updatedAt": "2026-05-12T16:00:00Z",
				"workspace": {
					"id": "workspace-id",
					"cwd": "/tmp/cmux",
					"git": {"branch": "feat/canonical"}
				},
				"activity": {"phase": "awaitingUser", "lastMessage": "Permission requested"},
				"attention": {"promptSummary": "Run tests"},
				"resume": {"command": ["codex", "resume", "canonical"]}
			}
		]
	}`))
	if err != nil {
		t.Fatalf("ParseState returned error: %v", err)
	}

	session := state.Sessions[0]
	if session.ResumeSessionID() != "canonical" {
		t.Fatalf("session id = %q, want canonical", session.ResumeSessionID())
	}
	if session.WorkingDir() != "/tmp/cmux" {
		t.Fatalf("cwd = %q, want /tmp/cmux", session.WorkingDir())
	}
	if session.Branch != "feat/canonical" {
		t.Fatalf("branch = %q, want feat/canonical", session.Branch)
	}
	if session.Preview != "Permission requested" {
		t.Fatalf("preview = %q, want Permission requested", session.Preview)
	}
	if session.ResumeCommand != "codex resume canonical" {
		t.Fatalf("resume command = %q, want codex resume canonical", session.ResumeCommand)
	}
}

File: tools/cmux-home/typescript/tests/state.test.ts
import { describe, expect, test } from "bun:test";
import { adapterCounts, groupSessionsByStatus, parseHomeState, statusCounts } from "../src/state";

describe("home state parsing", () => {
  test("parses direct sessions and normalizes statuses", () => {
    const state = parseHomeState({
      sessions: [
        {
          id: "one",
          agent: "Claude Code",
          session_id: "claude-1",
          status: "needs-input",
          title: "Approve plan",
          cwd: "/repo",
        },
        {
          id: "two",
          agent: "codex",
          sessionId: "codex-1",
          state: "in progress",
          summary: "Implement parser",
        },
      ],
    });

    expect(state.sessions).toHaveLength(2);
    expect(state.sessions[0]?.adapter).toBe("claude");
    expect(state.sessions[0]?.status).toBe("awaiting");
    expect(state.sessions[1]?.status).toBe("working");
  });

  test("loads workspace nested sessions with inherited cwd and branch", () => {
    const state = parseHomeState({
      workspaces: [
        {
          cwd: "/repo/cmux",
          branch: "feat-home",
          panels: [
            {
              id: "panel-1",
              terminal: {
                agent: {
                  kind: "opencode",
                  sessionId: "open-1",
                  status: "idle",
                  title: "Search OpenCode history",
                },
              },
            },
          ],
        },
      ],
    });

    expect(state.sessions).toHaveLength(1);
    expect(state.sessions[0]?.cwd).toBe("/repo/cmux");
    expect(state.sessions[0]?.branch).toBe("feat-home");
    expect(state.sessions[0]?.resumeCommand).toBe("cd /repo/cmux && opencode --session open-1");
  });

  test("groups and counts sessions by status and adapter", () => {
    const state = parseHomeState({
      sessions: [
        { agent: "pi", sessionId: "pi-1", status: "done", title: "Done" },
        { agent: "codex", sessionId: "codex-1", status: "running", title: "Run" },
        { agent: "claude", sessionId: "claude-1", status: "waiting", title: "Wait" },
      ],
    });

    expect(adapterCounts(state.sessions)).toEqual({ claude: 1, codex: 1, opencode: 0, pi: 1 });
    expect(statusCounts(state.sessions).working).toBe(1);
    expect(groupSessionsByStatus(state.sessions).map((group) => [group.status, group.sessions.length])).toEqual([
      ["awaiting", 1],
      ["working", 1],
      ["completed", 1],
    ]);
  });

  test("parses shared cmux home schema fields", () => {
    const state = parseHomeState({
      sessions: [
        {
          id: "codex:codex-permission-17",
          agent: "codex",
          agentSessionId: "codex-permission-17",
          title: "Fix auth callback race",
          status: "awaiting",
          summary: "Codex is waiting for a shell permission decision.",
          updatedAt: "2026-05-12T15:57:44Z",
          workspace: {
            cwd: "/Users/example/src/cmux",
            git: { branch: "feat-auth-callback" },
          },
          activity: {
            phase: "awaitingUser",
            lastMessage: "Permission request for test command.",
          },
          resume: {
            command: ["codex", "resume", "codex-permission-17"],
          },
        },
      ],
    });

    expect(state.sessions[0]?.status).toBe("awaiting");
    expect(state.sessions[0]?.sessionId).toBe("codex-permission-17");
    expect(state.sessions[0]?.cwd).toBe("/Users/example/src/cmux");
    expect(state.sessions[0]?.branch).toBe("feat-auth-callback");
    expect(state.sessions[0]?.preview).toBe("Permission request for test command.");
    expect(state.sessions[0]?.details).toBe("Codex is waiting for a shell permission decision.");
    expect(state.sessions[0]?.resumeCommand).toBe("codex resume codex-permission-17");
  });

  test("prefers canonical agentSessionId over legacy sessionId", () => {
    const state = parseHomeState({
      sessions: [
        {
          id: "codex:canonical",
          agent: "codex",
          sessionId: "legacy",
          agentSessionId: "canonical",
          title: "Canonical",
          status: "working",
        },
      ],
    });

    expect(state.sessions[0]?.sessionId).toBe("canonical");
  });

  test("rejects unsupported schema versions", () => {
    expect(() => parseHomeState({ schemaVersion: 2, sessions: [] })).toThrow(/schemaVersion/);
  });

  test("rejects schema statuses outside the shared contract", () => {
    expect(() => parseHomeState({
      schemaVersion: 1,
      sessions: [
        { id: "bad", agent: "codex", agentSessionId: "bad", status: "failed" },
      ],
    })).toThrow(/status/);
  });
});
